samples: openamp_rsc_table: enable support for imx95_evk

AndreHeinemans-NXP opened the pull request:
The openamp_rsc_table example now supports imx95_evk/mimx9596/m7. The A55 core must run linux-imx (tested with version 6.6.52 on BSP 2.2.0)

The example has been adapted to support both ipm and mbox drivers as messaging unit.

AndreHeinemans-NXP committed:
samples: openamp_rsc_table: enable support for imx95_evk

openamp_rsc_table will run on the i.MX95 M7 core and can communicate
to the A55 that runs linux-imx.

Signed-off-by: Andre Heinemans <andre.heinemans@nxp.com>

iuliana-prodan commented on soc/nxp/imx/imx9/imx95/m7/CMakeLists.txt:
```diff
@@ -5,4 +5,4 @@ zephyr_include_directories(.)
 zephyr_library_sources(soc.c)
 
-set(SOC_LINKER_SCRIPT ${ZEPHYR_BASE}/include/zephyr/arch/arm/cortex_m/scripts/linker.ld CACHE INTERNAL "")
+set(SOC_LINKER_SCRIPT ${CMAKE_CURRENT_SOURCE_DIR}/linker.ld CACHE INTERNAL "")
```
@AndreHeinemans-NXP you should split this commit in more patches.
Keep the i.MX95 M7 support for `openamp_rsc_table` sample separate that the changes to the sample.

For example, here's how I'm doing it: https://github.com/zephyrproject-rtos/zephyr/pull/79220/commits
I'm not saying is the best approach, but I believe is easier to review.

iuliana-prodan commented on samples/subsys/ipc/openamp_rsc_table/src/main_remote.c:
```diff
@@ -11,7 +11,15 @@
+#error Either CONFIG_IPM or CONFIG_MBOX must be enabled
+#endif
+
+#ifdef CONFIG_IPM
```
TBH I don't like this 
```
#ifdef
...
#else
..
#endif
```
There is a [driver to convert from mbox to ipm](https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/ipm/ipm_mbox.c), I'm using it for 8ULP (https://github.com/zephyrproject-rtos/zephyr/pull/83049/commits).

@arnopo @tnmysh what do you think?
IMO we should use the `mbox-to-ipm` driver since the IPM is not completely deprecated, and is still used in other samples.

iuliana-prodan commented on samples/subsys/ipc/openamp_rsc_table/Kconfig:
```diff
@@ -0,0 +1,13 @@
+
+source "Kconfig.zephyr"
+
+config COPY_RSC_TABLE
```
Why do you need this?
Why don't you put the resource table, from linker script, directly in the right location?
How to you load the firmware on M7?

dbaluta replied:
I think in this case it is loaded by the Linux running on the AP side.

dbaluta replied:
@AndreHeinemans-NXP  you should not use linux-imx but push people to work on upstreaming the Linux support. Until then you can use the NXP linux-nxp repo which is based to a close to upstream version.

AndreHeinemans-NXP replied:
@dbaluta, Yes I know. Right now we have to use linux-imx because mainline linux does not have the rsc_table defined in the dts. 

My guess is that mainline linux also requires system-manager to boot up the cores so copying the resource table in the m7 is probably still needed.

AndreHeinemans-NXP committed:
soc: nxp: imx: add .resource_table section for imx95

Add .resource_table section to the linker script for the
i.MX95. This section is used by intercore communication to
publish features and configurations to the remote

Signed-off-by: Andre Heinemans <andre.heinemans@nxp.com>

dbaluta commented on samples/subsys/ipc/openamp_rsc_table/src/main_remote.c:
```diff
@@ -155,7 +157,17 @@ int platform_init(void)
 	rsc_table_get(&rsc_table, &rsc_size);
+#endif
 	rsc_tab_physmap = (uintptr_t)rsc_table;
 
```
Please separate the main_remote.c chhange in a separate patch. As I said one patch per logical change. 

Then here lets factour out the operations of getting the rsc_table and create a separate function. 
```
void init_rsc_table(void) {
#ifdef CONFIG_COPY_RSC_TABLE
	{
		struct fw_resource_table *lrsc_table;

		rsc_table_get(&lrsc_table, &rsc_size);
		rsc_table = (struct fw_resource_table *)RSC_TABLE_ADDR;
		memcpy(rsc_table, lrsc_table, rsc_size);
	}

```
#else
	rsc_table_get(&rsc_table, &rsc_size);
#endif
}

AndreHeinemans-NXP replied:
Now while I am looking at it, shouldn't this function maybe replace the rsc_table_get() in lib/openamp/resource_table.c? So all applications can still do the rsc_table_get() as normal and don't need to take care about the copying of the table. The copying takes place in the rsc_table_get in case it is configured. There will be more applications and targets like the iMX95 who are booted the same way and can benefit from this feature.

@dbaluta What do you think about this?

AndreHeinemans-NXP committed:
lib: open-amp: optional copy of rsc_table to memory region

Option OPENAMP_COPY_RSC_TABLE is added for cases where the placement of the
.resource_table section is not controlled by the remote. This option lets
zephyr do the copying of the .resource_table section to a specified memory
region.

Signed-off-by: Andre Heinemans <andre.heinemans@nxp.com>

iuliana-prodan commented on lib/open-amp/resource_table.c:
```diff
@@ -29,6 +29,12 @@
 
+#ifdef CONFIG_OPENAMP_COPY_RSC_TABLE
+#define RSC_TABLE_ADDR	DT_REG_ADDR(DT_CHOSEN(zephyr_ipc_rsc_table))
+#define RSC_TABLE_SIZE	DT_REG_SIZE(DT_CHOSEN(zephyr_ipc_rsc_table))
```
why do you need the size?

AndreHeinemans-NXP replied:
Just to check if the table will fit the region. There is a BUILD_ASSERT on line 35

iuliana-prodan replied:
I saw that, but do you need a specific size for rsc_table or just to fill in the `reg` property?

AndreHeinemans-NXP replied:
By specifying a region with address and size, the build process will check for overlaps and will fail if there are any detected. This way it is guaranteed that copying the resource table won't write into any other memory region.

AndreHeinemans-NXP committed:
dts: nxp_imx95_m7: Add mu7 to dts

The Messaging Unit 7 peripheral is made available for
use with IPC such as OpenAMP/RpMsg. MU7 connects
to the A55 core

Signed-off-by: Andre Heinemans <andre.heinemans@nxp.com>

iuliana-prodan commented on samples/subsys/ipc/openamp_rsc_table/boards/imx95_evk_mimx9596_m7.overlay:
```diff
@@ -0,0 +1,39 @@
+		zephyr,ipc = &mailbox0;
+	};
+
+	soc {
```
`mu7` is already added in dtsi in soc section.
Just add it at the end of the file:
```
&mu7 {
	status = "okay";
};
```

dbaluta commented on samples/subsys/ipc/openamp_rsc_table/boards/imx95_evk_mimx9596_m7.conf:
```diff
@@ -0,0 +1,12 @@
+CONFIG_MBOX_INIT_PRIORITY=0
+CONFIG_LOG_MODE_MINIMAL=y
+CONFIG_LOG_DEFAULT_LEVEL=0
+CONFIG_OPENAMP_COPY_RSC_TABLE=y
```
This means that openamp_rsc_table is broken when used with remoteproc, right?

AndreHeinemans-NXP replied:
No because it is not possible to load the zephyr firmware by remoteproc. The boot process on the imx95 has been designed to have system-manager(m33) load the m7 firmware

dbaluta commented on doc/build/dts/api/api.rst:
```diff
@@ -422,6 +422,9 @@ device.
      - A node whose ``reg`` is used by the OpenAMP subsystem to determine the
        base address and size of the shared memory (SHM) usable for
```
We should avoid mentioning "remoteproc" in the commit message. Readers might not be aware how remoteproc loads the FW image. Even different implementations of remoteproc can load images in different ways.

AndreHeinemans-NXP replied:
I updated the explanation in lib/open-amp/Kconfig and the commit message

dbaluta commented on lib/open-amp/Kconfig:
```diff
@@ -34,3 +34,11 @@ config OPENAMP_RSC_TABLE_IPM_TX_ID
+	  If the binary is loaded in an other way than remoteproc then the
+	  resource_table section must be copied by the firmware to the correct
+	  location. This memory region can be declared in the device tree by the
+	  "zephyr,ipc_rsc_table" node
```
Similarly here, "binary is loaded in another way than remoteproc" doesn't offer too much of an explanation.
